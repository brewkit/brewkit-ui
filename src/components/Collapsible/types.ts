import { ReactNode } from 'react';


export interface Props {

    /**
     * Classes to be passed through to the component.
     */
    className?: string,

    /**
     * Static JSX, controls open/close functionality.
     */
    anchorContent: ReactNode,

    /**
     * Boolean to determine if collapse action should occur.
     */
    isCollapsed?: boolean,

    /**
     * Any custom content to fill the collapsible component.
     */
    children: ReactNode,

}
