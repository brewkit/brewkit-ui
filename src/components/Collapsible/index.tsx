import React, { ReactElement } from 'react';
import clsx from 'clsx';
import { Flipped, Flipper } from 'react-flip-toolkit';
import { Props } from './types';


/*
 ** Collapsible component uses CSS transitions to smoothly hide and show a block of content.
 */
function Collapsible({
    className,
    anchorContent,
    isCollapsed = true,
    children,
    ...otherProps
}: Props): ReactElement {


    const [collapsed, setCollapsed] = React.useState(isCollapsed);
    const wrapperClasses = clsx(
        'brew-Collapsible',
        { 'brew-Collapsible--open': !collapsed },
        className,
    );


    function toggleContent(): void {
        setCollapsed(!collapsed);
    }


    return (
        <div className={wrapperClasses} {...otherProps}>
            <Flipper flipKey={collapsed}>
                <div className="brew-Collapsible__controller" onClick={toggleContent}>
                    {anchorContent}
                </div>
                <Flipped flipId="children">
                    <div className="brew-Collapsible__content">
                        {children}
                    </div>
                </Flipped>
            </Flipper>
        </div>
    );

}


export default Collapsible;
