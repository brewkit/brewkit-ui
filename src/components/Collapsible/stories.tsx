import React, { ReactElement } from 'react';
import Collapsible from './index';
import Typography from '../Typography';
import Label from '../Label';
import { boolean } from '@storybook/addon-knobs';


export default {
    component: Collapsible,
    parameters: {
        componentSubtitle: <Label color="success">Stable</Label>,
    },
    title: 'Components|Display/Collapsible',
};


export const Default = (): ReactElement => {
    const isCollapsed = boolean('isCollapsed', true);
    const anchorContent = <Typography variant="h3">Click me to open</Typography>;

    return (
        <Collapsible anchorContent={anchorContent} isCollapsed={isCollapsed}>
            <Typography variant="body2">
                This is some content in a collapsible component.
            </Typography>
        </Collapsible>
    );
};
