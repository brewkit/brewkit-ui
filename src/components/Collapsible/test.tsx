import React from 'react';
import renderer from 'react-test-renderer';
import Collapsible from './index';
import { mount, shallow } from 'enzyme';


describe('Collapsible', () => {

    it('renders correctly', () => {
        const tree = renderer
            .create(<Collapsible anchorContent={(<div>Testing</div>)}>Test</Collapsible>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('mounts correctly', () => {
        const collapsible = shallow((
            <Collapsible
                anchorContent={(<div>Testing</div>)}
            >Test
            </Collapsible>
        ));
        expect(collapsible.hasClass('brew-Collapsible')).toBe(true);
    });

    it('correctly adds a className', () => {
        const className = `class-${Math.floor(Math.random() * 1000)}`;
        const collapsible = shallow((
            <Collapsible
                anchorContent={(<div>Testing</div>)}
                className={className}
            >Test
            </Collapsible>
        ));
        expect(collapsible.hasClass(className)).toBe(true);
    });

    it('correctly adds open class when isCollapsed is false', () => {
        const collapsible = mount((
            <Collapsible
                anchorContent={(<div>Testing</div>)}
                isCollapsed={false}
            >Test
            </Collapsible>
        ));
        expect(collapsible.prop('isCollapsed')).toBe(false);
        expect(collapsible.exists('.brew-Collapsible--open')).toBe(true);
    });

    it('renders children when passed in', () => {
        const wrapper = shallow((
            <Collapsible
                anchorContent={(<div>Testing</div>)}
            >
                <span>foo</span>
            </Collapsible>
        ));
        expect(wrapper.contains(<span>foo</span>)).toEqual(true);
    });

});
