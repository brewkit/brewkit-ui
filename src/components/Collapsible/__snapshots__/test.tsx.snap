// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible renders correctly 1`] = `
<div
  className="brew-Collapsible"
>
  <div>
    <div
      className="brew-Collapsible__controller"
      onClick={[Function]}
    >
      <div>
        Testing
      </div>
    </div>
    <div
      className="brew-Collapsible__content"
      data-flip-config="{\\"translate\\":true,\\"scale\\":true,\\"opacity\\":true}"
      data-flip-id="children"
      data-portal-key="portal"
    >
      Test
    </div>
  </div>
</div>
`;
